### PR TITLE
Introduce separate BlockByIdentifier class to get Layout Block based on CMS Block Identifier

### LINKED ISSUE #28309: [Issue] Introduce separate BlockByIdentifier class to get Layout Block based on CMS Block Identifier
This issue is automatically created based on existing pull request: magento/magento2#28147: Introduce separate BlockByIdentifier class to get Layout Block based on CMS Block Identifier

---------
### Preconditions: (*)

The existing class is anti-pattern:
1. Uses `\Magento\Framework\Model\AbstractModel::load` to load Model
1. If `block_id` is identifier, the `getIdentities` method returns same identity whatever the Scope is (`return [\Magento\Cms\Model\Block::CACHE_TAG . '_' . $this->getBlockId()];`), which is painful when you use the same `identifier` for multiple per-store CMS blocks.

Why can't we fix the original Block? Some of developers might have used:
```xml
        <referenceContainer name="page.top">
            <block class="Magento\Cms\Block\Block" name="header.info" after="navigation.sections">
                <arguments>
                    <argument name="block_id" xsi:type="string">666</argument>
                </arguments>
            </block>
        </referenceContainer>
```

That is why we keep old Block backwards-compatible and introducing new one that can be used explicitly:
```xml
        <referenceContainer name="page.top">
            <block class="Magento\Cms\Block\BlockByIdentifier" name="header.info" after="navigation.sections">
                <arguments>
                    <argument name="identifier" xsi:type="string">header-info</argument>
                </arguments>
            </block>
        </referenceContainer>
```

### Fixed Issues (if relevant)

1. magento/magento2#<issue_number>: Issue title

### Expected Result (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by lbajsarowicz**
### Description (*)

The existing class is anti-pattern:
1. Uses `\Magento\Framework\Model\AbstractModel::load` to load Model
1. If `block_id` is identifier, the `getIdentities` method returns same identity whatever the Scope is (`return [\Magento\Cms\Model\Block::CACHE_TAG . '_' . $this->getBlockId()];`), which is painful when you use the same `identifier` for multiple per-store CMS blocks.

Why can't we fix the original Block? Some of developers might have used:
```xml
        <referenceContainer name="page.top">
            <block class="Magento\Cms\Block\Block" name="header.info" after="navigation.sections">
                <arguments>
                    <argument name="block_id" xsi:type="string">666</argument>
                </arguments>
            </block>
        </referenceContainer>
```

That is why we keep old Block backwards-compatible and introducing new one that can be used explicitly:
```xml
        <referenceContainer name="page.top">
            <block class="Magento\Cms\Block\BlockByIdentifier" name="header.info" after="navigation.sections">
                <arguments>
                    <argument name="identifier" xsi:type="string">header-info</argument>
                </arguments>
            </block>
        </referenceContainer>
```

### Fixed Issues (if relevant)

1. magento/magento2#<issue_number>: Issue title

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#28309: Introduce separate BlockByIdentifier class to get Layout Block based on CMS Block Identifier

### CODE REVIEW

**Comment by lbajsarowicz** [comment]
After further investigation, I realized that additional Identities are not necessary, as these are used only as tags, not as cache keys. The only thing that was missing before and is missing now is to tag block with CMS Block identities, so that when CMS Block is requested to be removed from cache, the Layout Block cache should be cleaned too.

I'm going to adjust that now.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @lbajsarowicz,
Could you fix unit tests failures ?

**Comment by lbajsarowicz** [comment]
I can't find any failing report for Static Tests, neither for Unit Tests.

**Comment by ihor-sviziev** [comment]
Hi @slavvka could you add more info why unit and static tests fails? We don’t have anything in reports

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7560](https://jira.corp.magento.com/browse/ENGCOM-7560) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
MHI test fails due to used protected method, that was overridden from parent class, there is no way to fix it. I would say this is false positive that shouldn't require approval.

**Comment by slavvka** [comment]
@magento create issue

File: app/code/Magento/Cms/Block/BlockByIdentifier.php [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    /**
+     * Loads the CMS block by `identifier` provided as an argument
+     *
+     * @return BlockInterface|BlockModel
+     * @throws NoSuchEntityException
+     */
+    private function getCmsBlock(): BlockInterface
+    {
+        if (!$this->getIdentifier()) {
+            throw new NoSuchEntityException(
```
Comment (Serializator) [inline app/code/Magento/Cms/Block/BlockByIdentifier.php]: A `NoSuchEntityException` seems a bit unrelated to the actual cause of the exception, it wasn't that the entity was't found by the provided identifier, the identifier was simply not there.

Might there be a more suitable exception that could be thrown here?

File: app/code/Magento/Cms/Block/BlockByIdentifier.php [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    /**
+     * @inheritDoc
+     */
+    protected function _toHtml(): string
+    {
+        try {
+            return $this->filterOutput(
+                $this->getCmsBlock()->getContent()
+            );
+        } catch (NoSuchEntityException $e) {
```
Comment (Serializator) [inline app/code/Magento/Cms/Block/BlockByIdentifier.php:76]: If there is a block in the layout using the `Magento\Cms\Block\BlockByIdentifier` class with an invalid / unknown identifier it'd probably raise some eyebrows if this exception isn't logged somewhere.

It will most likely get trampled on by all the other logs but at the same time it might be a valuable clue for someone in the future.

File: app/code/Magento/Cms/Block/BlockByIdentifier.php (hunk 1 above)
Comment (lbajsarowicz) [reply app/code/Magento/Cms/Block/BlockByIdentifier.php]: Nice catch, this should actually be `InvalidArgumentException`

File: app/code/Magento/Cms/Block/BlockByIdentifier.php [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            ->filter($content);
+    }
+
+    /**
+     * Loads the CMS block by `identifier` provided as an argument
+     *
+     * @return BlockInterface|BlockModel
+     * @throws NoSuchEntityException
+     */
+    private function getCmsBlock(): BlockInterface
```
Comment (Serializator) [inline app/code/Magento/Cms/Block/BlockByIdentifier.php:112]: `private` prevents extensibility in a big way, rather use `protected` so that (for whatever reason) this class can be extended upon at least through a preference if really necessary. It's unfortunate that plugins can't be used on protected methods.

^ I won't comment on every private keyword in here (was thinking about it...) but I think you know what I'm grasping at!

File: app/code/Magento/Cms/Block/BlockByIdentifier.php (hunk 1 above)
Comment (Serializator) [reply app/code/Magento/Cms/Block/BlockByIdentifier.php]: I see what you did there! 😜

File: app/code/Magento/Cms/Block/BlockByIdentifier.php (hunk 2 above)
Comment (lbajsarowicz) [reply app/code/Magento/Cms/Block/BlockByIdentifier.php:76]: That was also actually something I was thinking about, as some of Administrators would like to have the block displayed conditionally - eg. for specific store, but don't have an error when the block by identifier is missing.

Also - logging a warning / notice for every single display of a page does not sound like good approach.

File: app/code/Magento/Cms/Block/BlockByIdentifier.php (hunk 3 above)
Comment (lbajsarowicz) [reply app/code/Magento/Cms/Block/BlockByIdentifier.php:112]: Using protected methods is discouraged and is going to fail Static code checks / Magento health check.

Also please consider reviewing these resources:
- https://devdocs.magento.com/guides/v2.3/ext-best-practices/extension-coding/common-programming-bp.html#composition-over-inheritance
- https://github.com/magento/magento2/issues/1019#issuecomment-72757989

So I am not going to change this into `protected` method as it's against what I consider as best practices.

File: app/code/Magento/Cms/Block/BlockByIdentifier.php (hunk 2 above)
Comment (Serializator) [reply app/code/Magento/Cms/Block/BlockByIdentifier.php:76]: I agree, it's a good point and didn't think about it that way 👍

File: app/code/Magento/Cms/Block/BlockByIdentifier.php [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        if (null === $this->cmsBlock) {
+            $this->cmsBlock = $this->blockByIdentifier->execute(
+                (string)$this->getIdentifier(),
+                $this->getCurrentStoreId()
+            );
+        }
+
+        return $this->cmsBlock;
```
Comment (Serializator) [inline app/code/Magento/Cms/Block/BlockByIdentifier.php:131]: What happens if the CMS block is disabled?

Comment (lbajsarowicz) [reply app/code/Magento/Cms/Block/BlockByIdentifier.php:131]: That is actually interesting catch, as

https://github.com/magento/magento2/blob/c74256b9c637ac9808f8b0877d895b4a9b44f90f/app/code/Magento/Cms/Model/GetBlockByIdentifier.php#L43-L54

does not care about Block status. I'm going to adjust the logic.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run Unit Tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@slavvka could you explain why DB compare test fails?

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved!

Failed functional test isn't related to changes from this PR, failed MHI looks like false positive.

**Comment by lbajsarowicz** [comment]
@magento run all tests

I'd like to verify if there are no issues with current mainline.

**Comment by engcom-Echo** [comment]
:heavy_check_mark: QA Passed
A simple block was created, using class = "Magento \ Cms \ Block \ BlockByIdentifier". getIdentities() return different result for different store. Before, the function returned the same result.

**Comment by ihor-sviziev** [comment]
@engcom-Alfa @engcom-Echo is there any reasons why this PR isn't moving forward?

**Comment by engcom-Foxtrot** [comment]
@magento create issue

**Comment by engcom-Foxtrot** [comment]
Magento Health Index failure should be approved.

**Comment by ihor-sviziev** [comment]
@engcom-Foxtrot MHI test failure is false positive! 
It fails on the method https://github.com/magento/magento2/pull/28147/files#diff-014e853ca19a6acdd6c6c6a51aa6fb7eR70
```php
protected function _toHtml(): string
``
which is overrides protected method from parent.

**Comment by sidolov** [comment]
Hi @ihor-sviziev! All failures should be approved, even if they are false-positive to make sure we will not introduce breaking changes. We will proceed with the approval ASAP

**Comment by lbajsarowicz** [comment]
@sidolov Ping.

**Comment by gabrieldagama** [comment]
Internal approval has been created and being processed (MC-37066).

**Comment by lbajsarowicz** [comment]
@gabrieldagama Where we are regarding that PR?

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-7560](https://jira.corp.magento.com/browse/ENGCOM-7560) has been created to process this Pull Request

File: app/code/Magento/Cms/Block/Block.php [hunk 5]
```diff
@@ -10,6 +10,8 @@
 
 /**
  * Cms block content block
+ * @deprecated This class introduces caching issues and should no longer be used
```
Comment (vovayatsyuk) [inline app/code/Magento/Cms/Block/Block.php:13]: What issues?

Comment (vovayatsyuk) [reply app/code/Magento/Cms/Block/Block.php:13]: Must be this one, that was fixed? https://github.com/magento/magento2/commit/5c84c1ad6aecbf6209d04bcaa160c49f356344d4

File: app/code/Magento/Cms/Block/BlockByIdentifier.php [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    /**
+     * Returns the value of `identifier` injected in `<block>` definition
+     *
+     * @return string|null
+     */
+    private function getIdentifier(): ?string
+    {
+        return $this->getdata('identifier') ?: null;
```
Comment (vovayatsyuk) [inline app/code/Magento/Cms/Block/BlockByIdentifier.php]: Must be getData

**Review by vovayatsyuk — CHANGES_REQUESTED** [review]

File: app/code/Magento/Cms/Block/BlockByIdentifier.php (hunk 6 above)
Comment (ihor-sviziev) [reply app/code/Magento/Cms/Block/BlockByIdentifier.php]:
```suggestion
        return $this->getData('identifier') ?: null;
```

**Comment by vovayatsyuk** [comment]
I don't understand why do Magento needs this new block. It seems that everything from this PR could be done in the existing block without creating a new entity and without BC.

Something like this:

```php
if (is_numeric($identifier)) {
  // loadById to be compatible with custom code that uses id.
} else {
 // new logic
}
```

File: app/code/Magento/Cms/Block/Block.php (hunk 5 above)
Comment (lbajsarowicz) [reply app/code/Magento/Cms/Block/Block.php:13]: Not really. Except the fact the previous one uses both `id` and `identifier` for cache key, there was something more (I'll try to find out what was the other issue)

**Comment by lbajsarowicz** [comment]
@vovayatsyuk If we want to follow SOLID principles - the arguments have to explicit, without type juggling. It's exactly the same like Page by Identifier interface.

**Comment by vovayatsyuk** [comment]
But we already have a block to show cms blocks. I can only agree with this change if we'll remove another block in the future. But this leads to the next thought - why create a new block instead of refactoring the existing one and remove deprecated logic in the next major release?

**Comment by lbajsarowicz** [comment]
@vovayatsyuk The only issue for introducing that Interface was the requirement for backwards compatibility, as this PR was created **before 2.4 was released**. If I knew that's going to hang for a year, I'd just modify the original implementation with happy "do it all" (against the SOLID principles) way.

**Comment by ihor-sviziev** [comment]
@vovayatsyuk in my opinion - we can go with both options. in the current implementation of cms/block/block we have following issues:
1. not enforced to use CMS block ID or Identified, and both currently supported and widely used by extension developers
2. different cache tags in case if block ID or identified was used
3. deprecated method of loading the block (using $model->load())
4. In case if the block with such identifier / id is missing / disabled  / not assigned to current store view - we don't have any errors/warnings/etc.

Also as @lbajsarowicz said - addressing all of these issues will cause the not following SOLID principles and probably will be backward incompatible.
Technically this block isn't marked as `@api` and we might add backward incompatible changes, but if we'll change the logic - it definitely will cause issues during magento upgrade.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@sidolov is there any reason why this PR isn't getting merged?

**Comment by sidolov** [comment]
@ihor-sviziev the pull request in the delivery queue right now and will be delivered soon. There are no reasons to postpone the merging process.

**Comment by lbajsarowicz** [comment]
@sidolov , @ihor-sviziev @vovayatsyuk - Thanks for your support. Glad to have this merged.
